src/matrixrtc/NewMembershipManager.ts

```diff
@@ -0,0 +1,808 @@
+                if (state.delayId) {
+                    // Remove all running updates and restarts
+                    this.scheduler.resetActions([]);
+                    try {
+                        await this.client._unstable_updateDelayedEvent(state.delayId, UpdateDelayedEventAction.Cancel);
+                        state.delayId = undefined;
+                        state.updateDelayedEventRetries = 0;
+                        this.scheduler.addAction({
+                            ts: Date.now(),
+                            type: DelayedLeaveActionType.SendFirstDelayedEvent,
+                        });
+                    } catch (e) {
+                        this.handleRateLimitError(
+                            e,
+                            state.updateDelayedEventRetries,
+                            (retryIn) => {
+                                state.updateDelayedEventRetries++;
+                                this.scheduler.addAction({
+                                    ts: Date.now() + retryIn,
+                                    type: DelayedLeaveActionType.SendFirstDelayedEvent,
+                                });
+                            },
+                            () => {
+                                throw Error(
+                                    "Exceeded maximum delayed event update (cancel) attempts (client._unstable_updateDelayedEvent): " +
+                                        e,
+                                );
+                            },
+                        );
+                        this.handleNotFoundError(e, () => {
+                            // If we get a M_NOT_FOUND we know that the delayed event got already removed.
+                            // This means we are good and can set it to undefined and run this again.
+                            state.delayId = undefined;
+                            this.scheduler.addAction({
+                                ts: Date.now(),
+                                type: DelayedLeaveActionType.SendFirstDelayedEvent,
+                            });
+                        });
+                    }
+                } else {
+                    try {
+                        const response = await this.client._unstable_sendDelayedStateEvent(
+                            this.room.roomId,
+                            {
+                                delay: this.membershipServerSideExpiryTimeout,
+                            },
+                            EventType.GroupCallMemberPrefix,
+                            {}, // leave event
+                            this.stateKey,
+                        );
+                        // Success we reset retires and set delayId.
+                        state.sendDelayedEventRetries = 0;
+                        state.delayId = response.delay_id;
+                        this.scheduler.addAction({ ts: Date.now(), type: MembershipActionType.SendJoinEvent });
```

Comment:
I *think* this is the only place where the `SendJoinEvent` action is scheduled. So, what happens if the server doesn't support the delayed events feature? At the moment I don't *think* `SendJoinEvent` gets run.

Reply:
yes I really need to go through all catch blocks and think about the most reasonable "default" failure mode.
Asking the question: "Can we do the same thing in ALL other failure cases and what would that message be?"
I think that will expose a couple of other specific errors we need to handle specifically and what we need to do in the "default" case.

Reply:
I went through all the catch blocks and opted for throwing in a lot of cases.
Only on the actual event sending I opted for a blind "retry" even though we do not understand the error.
But this blind retry is now also rate limited.

Reply:
This got another update in 5a680b0e9fa43024d93fbad9f675d44b6c0585ac where I added tests for those cases